---
title: "20231016_SDR005_MB_check_overlap"
author: "Dominik Lindenhofer"
date: "2023-10-16"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Load libs


```{r}


#Load Primer3

Sys.setenv(PATH = paste("/Users/dominik.lindenhofer/opt/miniconda3/bin", Sys.getenv("PATH"), sep = ":"))
Sys.setenv(PATH = paste("/Users/dominik.lindenhofer/Packages/mosaic/mosaic/bin", Sys.getenv("PATH"), sep = ":"))


#Load Blastn
Sys.setenv(PATH = paste("/Users/dominik.lindenhofer/ncbi-blast-2.13.0+/bin", Sys.getenv("PATH"), sep = ":"))


library(TAPseq)
library(GenomicRanges)
library(BiocParallel)
library(BSgenome)
library(rtracklayer)
library(dplyr)
library(biomaRt)
library(ggplot2)
library(BSgenome.Mmusculus.UCSC.mm10)

```

#Load list of genes to run

```{r}

TAP_input_df <- read.csv("/Users/dominik.lindenhofer/ownCloud/Shared_project_folder/scDNA-scRNA-seq/20250124_SDR009_Cell_Mixing/01_Gene_sel/Plots/3T3_genes_selected_HML.csv", header = TRUE)

```



## Run TAP-seq pipeline

```{r}



# human genome BSgenome object (needs to be istalled from Bioconductor)
mm10 <- getBSgenome("BSgenome.Mmusculus.UCSC.mm10")

# download and import gencode hg38 annotations
#url <- "ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_40/gencode.v40.annotation.gtf.gz"
#url <- "ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_40/gencode.v40.annotation.gtf.gz"
#annot <- import(url, format = "gtf")

annot <- rtracklayer::import('/Users/dominik.lindenhofer/Documents/Data_Local/TAP_seq_gencode.vM25/gencode.vM25.annotation.gtf.gz')

#annot_df <- annot %>% as.data.frame() %>% tibble::as_tibble() 

#all_gene_names <- unique(annot_df$gene_name)
target_gene_names <- as.character(TAP_input_df$Gene)

#target_genes_O <- intersect(target_gene_names, element_names)
#target_genes_NO <- setdiff(target_gene_names, element_names)

#TAP_input_df <- TAP_input_df %>% mutate(Gene_found = case_when(Gene %in% target_genes_O ~ "Found", Gene %in% target_genes_NO ~ "Not_Found"))
                                                              


# write.table(TAP_input_df,file="/Users/dominik.lindenhofer/ownCloud/Shared_project_folder/scDNA-scRNA-seq/20231006_SDR005_BCL_pilot/2_TAP_seq/20231016_SDR005_TAP_seq_input_mod_v2.tsv", quote = FALSE, row.names = FALSE, col.names = TRUE, sep = "\t")



# extract exon annotations for protein-coding genes to build transcripts
#tx_exons <- annot[annot$type == "exon" & annot$gene_type == "protein_coding"]

tx_exons <- annot[annot$type == "exon"]

# Names of target genes
target_gene_names <- as.character(TAP_input_df$Gene)

#remove non working ones 
#target_gene_names <- target_gene_names[target_gene_names != "IGLC6"]


#Select subset to run pipeline 
#target_gene_names <- target_gene_names[1]

target_genes <- tx_exons[tx_exons$gene_name %in% target_gene_names]
target_genes <- split(target_genes, f = target_genes$gene_name)


element_names <- names(target_genes)

#Rename chromosome names to fit bam file

chr_names <- target_genes@unlistData@seqnames@values
chr_all <- levels(chr_names )

chr_names <- paste0("mm10___", chr_names)
chr_all <- paste0("mm10___", chr_all)
chr_names <- factor(chr_names, levels = chr_all )

target_genes@unlistData@seqnames@values <- chr_names 

target_genes@unlistData@seqinfo@seqnames <- chr_all


print('target genes:')
print(target_genes)




#BAM file downloaded from here - https://www.10xgenomics.com/datasets/500-1-1-mixture-of-human-hek-293-t-and-mouse-nih-3-t-3-cells-3-lt-v-3-1-chromium-x-3-1-low-6-1-0

#smaller bam file used
scRNAseq_bam <- "500_hgmm_3p_LT_Chromium_X_possorted_genome_bam.bam"


# register backend for parallelization
register(MulticoreParam(workers = 10))

print('inferring polyA sites...')

# infer polyA sites from Drop-seq data
polyA_sites <- inferPolyASites(target_genes, bam = scRNAseq_bam, polyA_downstream = 50,
                               wdsize = 100, min_cvrg = 1, parallel = TRUE)



# In a real-world application it's adviced to
# manually inspect the polyA site predictions and remove any obvious false positives. This is easiest
# done by exporting the polyA sites into a .bed. This file can then be loaded into a genome browser
# (e.g. IGV) together with the target gene annotations and the .bam file to visually inspect the polyA
# site predictions.

#Convert chr names back in pA sites and target genes

chr_names <- polyA_sites@seqnames@values
chr_all <- levels(chr_names )

chr_names <- gsub("mm10___", "", chr_names)
chr_all <- gsub("mm10___", "", chr_all)
chr_names <- factor(chr_names, levels = chr_all )


polyA_sites@seqnames@values <- chr_names
polyA_sites@seqinfo@seqnames <- chr_all

#Simply load target genes again

target_genes <- tx_exons[tx_exons$gene_name %in% target_gene_names]
target_genes <- split(target_genes, f = target_genes$gene_name)



export(polyA_sites, con = "./3T3_polyA_sites.bed", format = "bed")
export(unlist(target_genes), con = "./3T3_target_genes.gtf", format = "gtf")


# Once we are happy with our polyA predictions, we then use them to truncate the target gene
# transcripts. All target gene transcripts overlapping any polyA sites are truncated at the polyA
# sites, as we assume that they mark the transcripts 3' ends. By default the most downstream polyA
# site is taken to truncate a transcript. If multiple transcripts per gene overlap polyA sites, a
# merged truncated transcript model is generated from all overlapping transcripts. If no transcript
# of a given gene overlap with any polyA sites a consensus transcript model is created from all
# transcripts for that gene.

print('truncating transcripts...')

# truncate transcripts at inferred polyA sites
truncated_txs <- truncateTxsPolyA(target_genes, polyA_sites = polyA_sites, parallel = TRUE)

# To finalize this part, we need to extract the sequences of the truncated transcripts. Here we use
# Bioconductor's BSgenome to export transcript sequences, but the genome sequence could also be
# imported from a fasta file into a DNAStringSet object.



#Change them back 



# get sequence for all truncated transcripts
txs_seqs <- getTxsSeq(truncated_txs, genome = mm10)


# Design  primers for MB

# MODIFICATION - only one primer predicted as there is no nested PCR done

# Primer3 uses boulder-IO records for input and output (see: http://primer3.org/manual.html). TAPseq
# implements TsIO and TsIOList objects, which store Primer3 input and output for TAP-seq primer
# design in R's S4 class system. They serve as the users interface to Primer3 during primer design.

# First we create Primer3 input for outer forward primer design based on the obtained transcript
# sequences. The reverse primer used in all PCR reactions is used, so that Primer3 can pick forward
# primers suitable to work with it. By default this function uses the 10x Beads-oligo-dT and right
# primer, and chooses optimal, minimum and maximum melting temperatures based on that. If another
# protocol (e.g. Drop-seq) is used, these parameters can be adapted.

print('creating inner primers...')


# create new TsIO objects for inner primers, note the different product size
inner_primers <- TAPseqInput(txs_seqs, target_annot = truncated_txs,
                             product_size_range = c(150, 300), primer_num_return = 5,
                             primer_opt_tm = 60,
                             primer_min_tm = 58,
                             primer_max_tm = 62)



# design inner primers
inner_primers <- designPrimers(inner_primers)

#Blast primers
print('creating BLASTDb')

# create blast database
blastdb <- file.path(tempdir(), "blastdb")
createBLASTDb(genome = mm10, annot = tx_exons, blastdb = blastdb)

help(createBLASTDb)
print('BLASTING inner primers')



print('BLASTing inner primers')

# blast inner primers
inner_primers <- blastPrimers(inner_primers, blastdb = blastdb, max_mismatch = 0,
                              min_aligned = 0.75)

print('picking best targets')
                              
# pick best primer per target gene
best_inner_primers <- pickPrimers(inner_primers, n = 1, by = "off_targets")



inner_comp <- checkPrimers(best_inner_primers)


comp <- bind_rows(inner = inner_comp, .id = "set")

# add variable for pairs with any complemetarity score higher than 47
comp <- comp %>%
  mutate(high_compl = if_else(primer_pair_compl_any_th > 47 | primer_pair_compl_end_th > 47,
                              true = "high", false = "ok")) %>% 
  mutate(high_compl = factor(high_compl, levels = c("ok", "high")))

print('complementarity scores:')
print(comp)


# plot complementarity scores
ggplot(comp, aes(x = primer_pair_compl_any_th, y = primer_pair_compl_end_th)) +
  facet_wrap(~set, ncol = 2) +
  geom_point(aes(color = high_compl), alpha = 0.25) +
  scale_color_manual(values = c("black", "red"), drop = FALSE) +
  geom_hline(aes(yintercept = 47), colour = "darkgray", linetype = "dashed") +
  geom_vline(aes(xintercept = 47), colour = "darkgray", linetype = "dashed") +
  labs(title = "Complementarity scores TAP-seq primer combinations",
       color = "Complementarity") +
  theme_bw()




# create data.frames for outer and inner primers
inner_primers_df <- primerDataFrame(best_inner_primers)

write.csv(inner_primers_df,'SDR009_3T3_best_inner_primers.csv')


# Furthermore we can create BED tracks with the genomic coordinates of the primer binding sites for
# viewing in a genome browser.

# create BED tracks for outer and inner primers with custom colors
inner_primers_track <- createPrimerTrack(best_inner_primers, color = "goldenrod1")

# the output data.frames contain lines in BED format for every primer
head(inner_primers_track)

# export tracks to .bed files ("" writes to the standard output, replace by a file name)
exportPrimerTrack(inner_primers_track, con = "./SDR009_inner_primers.bed")


```




#Further analysis 

```{r}

#Integrate the expression values

data_exp <- read.csv("/Users/dominik.lindenhofer/ownCloud/Shared_project_folder/scDNA-scRNA-seq/20250124_SDR009_Cell_Mixing/01_Gene_sel/Plots/3T3_genes_selected_HML.csv")

inner_primers_df <- inner_primers_df %>% left_join(data_exp, by = c("seq_id" = "Gene"))

write.csv(inner_primers_df,'SDR009_3T3_best_inner_primers.csv')
```
