---
title: "scDNA-RNA Maturation Analysis"
author: "Donnacha Fitzgerald"
date: "2023-11-28"
output: html_document
---

The purpose of this script is to analyze B-cell maturation state and variant composition of FL (FL1, FL2) and DLBCL (GCB1) samples from scDNA-RNA-seq data generated by Dominik Lindenhofer. It follows the scripts: SDR005_data_share/20240205_SDR005_SDRranger_gDNA-V4_cDNA-V5/20240205_SDR005_Run1_SDRranger.Rmd.
SDR005_data_share/20240205_SDR005_SDRranger_gDNA-V4_cDNA-V5/20240205_SDR005_Run2_SDRranger.Rmd.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(tidyverse)
library(Seurat)
set.seed(23)
```

# Load Data
```{r}
# Load QC filtered Seurat object from all 3 samples. 
sobj <- readRDS("/g/huber/users/fitzgerald/R/scDNA-RNA/Shared_Dominik/SDR005_data_share/20240205_SDR005_SDRranger_gDNA-V4_cDNA-V5/SDR005_Combined_Geno_full.rds")
```

# RNA-Seq Analysis

## Preprocessing

```{r}
# Log normalization
sobj <- NormalizeData(sobj)

# Find variable features
sobj <- FindVariableFeatures(sobj, selection.method = "vst")

# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(sobj), 10)

# plot variable features with and without labels
plot1 <- VariableFeaturePlot(sobj)
LabelPoints(plot = plot1, points = top10, repel = TRUE)  + 
  ggtitle("Gene Expression Variance") +
  NoLegend()
ggsave("Shared_Dominik/SDR005_data_share/20240205_SDR005_SDRranger_gDNA-V4_cDNA-V5/Supp7/VariableFeatures.pdf", width = 4, height = 4)

# Scaling
all.genes <- rownames(sobj)
sobj <- ScaleData(sobj, features = all.genes)
```


```{r}
# PCA
sobj <- FindVariableFeatures(sobj, selection.method = "vst")
sobj <- RunPCA(sobj, features = VariableFeatures(object = sobj))
VizDimLoadings(sobj, dims = 1:2, reduction = "pca")
DimHeatmap(sobj, dims = 1:15, cells = 500, balanced = TRUE)
DimPlot(sobj, reduction = "pca", shuffle = TRUE)
```

```{r}
ElbowPlot(sobj)
```

Little variance is captured beyond 10 PCs.

```{r}
# UMAP
sobj <- RunUMAP(sobj, dims = 1:10)
```

## Sample Annotation

```{r}
# Add sample labels
Idents(sobj) <- "Type"
new.cluster.ids <- c("FL1", "FL1", "GCB1", "GCB1", "GCB1", "FL1", "GCB1", "FL1", "FL2", "FL2", "FL2", "FL2", "FL2", "FL2", "FL2", "FL2")
names(new.cluster.ids) <- levels(sobj)
sobj <- RenameIdents(sobj, new.cluster.ids)
sobj$sample <- Idents(sobj)

# Set sample color palette
sample_colors <- c("#F2673C", "#F8AFB0", "#182F57")

DimPlot(sobj, group.by = "sample", reduction = "umap", label = FALSE, pt.size = 0.1, shuffle = TRUE, cols = sample_colors) + 
  coord_fixed() + 
  labs(x = "UMAP 1", y = "UMAP 2", title = "Samples")  +
  theme(legend.position = "bottom")
ggsave("Shared_Dominik/SDR005_data_share/20240205_SDR005_SDRranger_gDNA-V4_cDNA-V5/Supp7/SampleUMAP.pdf", width = 4, height = 5)
```

## Entity Annotation 

```{r}
# Add entity labels
new.cluster.ids <- c("FL", "DLBCL", "FL")
names(new.cluster.ids) <- levels(sobj)
sobj <- RenameIdents(sobj, new.cluster.ids)
sobj$entity <- Idents(sobj)
DimPlot(sobj, reduction = "umap", label = TRUE, pt.size = 0.5, shuffle = TRUE) + coord_fixed() + NoLegend()
```

## Clustering
```{r}
sobj <- FindNeighbors(sobj, dims = 1:10)
## Resolution = 0.5
sobj <- FindClusters(sobj, resolution = 0.5)
sobj$clusters_0.5 <- Idents(sobj)
DimPlot(sobj)
```

```{r}
# Resolution = 1
sobj <- FindClusters(sobj, resolution = 1)
sobj$clusters_1 <- Idents(sobj)
DimPlot(sobj)
```

```{r}
# find markers for every cluster compared to all remaining cells, report only the positive
# ones
Idents(sobj) <- sobj$clusters_0.5
Markers <- FindAllMarkers(sobj, only.pos = TRUE)
Markers %>%
    group_by(cluster) %>%
    dplyr::filter(avg_log2FC > 1)
```

```{r, fig.width=7, fig.height=10}
Markers %>%
    group_by(cluster) %>%
    dplyr::filter(avg_log2FC > 1) %>%
    slice_head(n = 10) %>%
    ungroup() -> top10
DoHeatmap(sobj, features = top10$gene) + NoLegend()
```

## Identify B Cells

```{r}
FeaturePlot(sobj, features = c("IGKC", "IGLC2", "PAX5", "MS4A1"))
```


0: Other
1: B
2: B
3: B
4: B
5: B
6: B
7: B
8: B
9: B
10: Other
11: B



```{r}
Idents(sobj) <- sobj$clusters_0.5
DimPlot(sobj, label = TRUE)
new.cluster.ids <- c("Other", "B", "B", "B", "B", "B", "B", "B", "B", "B", "Other", "B")
names(new.cluster.ids) <- levels(sobj)
sobj <- RenameIdents(sobj, new.cluster.ids)
sobj$B <- Idents(sobj)
DimPlot(sobj, reduction = "umap", label = TRUE, pt.size = 0.5) + NoLegend()
```

```{r}
# Isolate B-cells
B <- subset(sobj, B == "B")
Idents(B) <- B$sample
DimPlot(B) + coord_fixed()
```

## Identify Malignant cells

Malignant populations are light-chain restricted (ie. all cells either kappa or lambda positive). In the CITE-Seq dataset, this FL sample (FL1) is lambda-restricted while the DLBCl sample is kappa-restricted. Both samples consist of >90% malignant B-cells.

Although we don't have surface protein light chain abundances, we can use the transcripts as a surrogate.

```{r}
# Calculate kappa light chain proportion
LC <- FetchData(B, vars = c("IGKC", "IGLC1", "IGLC2", "IGLC3", "IGLC6", "IGLC7"), slot = "counts")
B$KLR <- LC$IGKC/(LC$IGLC2 + LC$IGLC3 + LC$IGLC7 + LC$IGKC)
# replace NAs, in which no light chain was detected, with 0.5
B$KLR[is.na(B$KLR)] <- 0.5
VlnPlot(B, features = c("IGKC", "IGLC2", "KLR"), pt.size=0.1) + 
  stat_summary(fun = "mean",
               geom = "crossbar", 
               width = 0.5,
               colour = "blue")
```

FL1 appears to be lambda-restricted, FL2 kappa-restricted, while GCB1 appears to have light-chain depletion.

```{r, fig.width = 2.8, fig.height = 5}
FeaturePlot(B, features = c("IGKC", "IGLC2"), split.by = "sample", pt.size=0.1)
FeaturePlot(B, features = "KLR", split.by = "sample", pt.size=0.1, cols = c("blue", "yellow", "red"), by.col = FALSE)
ggsave("Shared_Dominik/SDR005_data_share/20240205_SDR005_SDRranger_gDNA-V4_cDNA-V5/Supp7/KLR_UMAPs.pdf", width = 2.8, height = 5)
```


```{r}
VlnPlot(B, features = "KLR", group.by = "clusters_1") + stat_summary(fun = mean, geom='point', size = 10, colour = "green", shape = 95)+
  geom_hline(yintercept = c(0.25, 0.75), col= c("blue","red")) +
  labs(x = "Samples", y = "Kappa Light Chain Propotion") +
  ggtitle("Light Chain Restriction by Tumor") +
  theme(text = element_text(size = 7),
        plot.title = element_text(size = 7, face = "bold"),
        axis.text = element_text(size = 5))
```
Almost all clusters appear to be light-chain restricted. There may be a difference in sensitivity between both markers, making it difficult to identify non-malignant B-cells confidently in FL1.

## Maturation state marker profiling
```{r}
# select maturation features from FeatureSelection.Rmd
features <- c("IGHD", "IGHM", "TCL1A", "SELL", "CD1C", "CXCR5", "MME", "BCL6", "CCNB1", "AICDA", "CAMK1", "CD72", "MS4A1", "PTPN6", "SLA", "FCRL2", "CFLAR", "FOXP1", "CD83", "EBI3", "BTK", "BLK", "CD74", "BLNK", "CD40", "NFKB1", "NFKB2", "TRAF1", "ICAM1", "REL", "RELB", "BACH2", "CCR6", "GPR183", "TNFRSF17", "IRF4", "PRDM1", "CD38", "FAS", "IGHG1", "IGHA1", "IGHE", "CXCR3", "IGLC2", "IGKC")
```

```{r, fig.width=7, fig.height=7}
Idents(B) <- B$clusters_1
DimPlot(B) + coord_fixed()
DotPlot(B, features = rev(features), dot.scale = 4) + coord_flip()
```

## Maturation State Annotation

Annotation of B-cell maturation states from a reactive lymph node reference (Fitzgerald et al, 2023).

```{r}
# Load annotated single-cell reference from reactive lymph nodes
rLN <- readRDS("/g/huber/users/fitzgerald/R/Workflowr/cite-seq/data/Objects/rLN.rds")
DefaultAssay(rLN) <- "RNA"
```

```{r}
# Split samples
Tumors <- SplitObject(B, split.by = "sample")
# Normalize query data to match the rLN reference normalization
Tumors <- lapply(Tumors, NormalizeData, normalization.method = "LogNormalize", 
                        scale.factor = 10000)
```


```{r, message=FALSE, verbose=FALSE}
set.seed(23)
# Reference mapping function
refmap <- function(x){
  anchors <- FindTransferAnchors(reference = rLN, query = x,
    dims = 1:50, reference.reduction = "pca")
  x <- MapQuery(anchorset = anchors, reference = rLN, query = x,
    refdata = list(State = "State"), reference.reduction = "pca")
  x$predicted.State <- factor(x$predicted.State, 
                      levels = c("NaÃ¯ve", "DZ", "LZ", "Mem IgM", "Mem IgG", "Plasma"))
  return(x)
}
Tumors <- lapply(Tumors, refmap)
```

## Visualize mapping
```{r}
# Set color palette for maturation states
mathue9 <- c("#ff707c",
              "#975F24",
              "#f8d058",
              "#41d5a8",
              "#006c0b",
              "#902395")
# Add color for non-B cells (Other)
mathue10 <- c("#ff707c",
              "#975F24",
              "#f8d058",
              "#41d5a8",
              "#006c0b",
              "#902395",
              "#808080")

lapply(Tumors, DimPlot, group.by = "predicted.State",
       label = TRUE, repel = TRUE,
        label.box = TRUE, 
        label.color = "white",
        label.size = 4,
       reduction = "umap",
       cols = mathue9)
lapply(Tumors, FeaturePlot, features = "predicted.State.score", cols = c("blue", "red"))
lapply(Tumors, VlnPlot, features = "predicted.State.score", group.by = "predicted.State", cols = mathue9, alpha = 0)
```

DZ state is more confidently assigned than the other states, followed by LZ.

```{r, fig.height=7}
DotPlot(Tumors$FL1, features = rev(features), dot.scale = 4, group.by = "predicted.State") + coord_flip()
DotPlot(Tumors$FL2, features = rev(features), dot.scale = 4, group.by = "predicted.State") + coord_flip()
DotPlot(Tumors$GCB1, features = rev(features), dot.scale = 4, group.by = "predicted.State") + coord_flip()
```
DZ and LZ markers correspond well, although other states are difficult to assign with confidence.

```{r}
Tumors_merged <- merge(Tumors$FL1, y = c(Tumors$FL2, Tumors$GCB1))
```

```{r}
# Label all states except DZ and LZ as "B other"
Idents(Tumors_merged) <- "predicted.State"
new.cluster.ids <- c("LZ", "B unassigned", "DZ", "B unassigned", "B unassigned", "B unassigned")
names(new.cluster.ids) <- levels(Tumors_merged)
Tumors_merged <- RenameIdents(Tumors_merged, new.cluster.ids)
Tumors_merged$predicted.State2 <- Idents(Tumors_merged)
```

```{r}
Tumors_merged[["predicted.State3"]] <- Tumors_merged[["predicted.State"]]

# Access metadata
metadata <- Tumors_merged@meta.data

# Assign "B unassigned" based on condition
metadata$predicted.State3[metadata$predicted.State.score < 0.75] <- "B unassigned"

# Update the metadata in your Seurat object
Tumors_merged@meta.data <- metadata
```

```{r, eval = FALSE}
# Save sample objects
saveRDS(Tumors_merged, "/g/huber/users/fitzgerald/R/scDNA-RNA/Shared_Dominik/SDR005_data_share/20240205_SDR005_SDRranger_gDNA-V4_cDNA-V5/SDR005_Seurat_Bcells_annotated.rds")
```

## Plot Light Chain Restriction
```{r}
Idents(Tumors_merged) <- "sample"
VlnPlot(Tumors_merged, features = c("IGKC", "IGLC2", "KLR"), pt.size=0.1) + 
  stat_summary(fun = "mean",
               geom = "crossbar", 
               width = 0.5,
               colour = "blue")
```

FL1 appears to be lambda-restricted, FL2 kappa-restricted, while GCB1 appears to have light-chain depletion.

```{r, fig.height=7, fig.width=5}
DotPlot(Tumors_merged, features = rev(features), dot.scale = 4, group.by = "predicted.State") + coord_flip() + RotatedAxis()
```

```{r}
DimPlot(B, group.by = "CellTypes", split.by = "sample", pt.size=0.1, cols = mathue10, ncol = 1)
ggsave("Shared_Dominik/SDR005_data_share/20240205_SDR005_SDRranger_gDNA-V4_cDNA-V5/Supp7/State_UMAPs.pdf", width = 3.5, height = 5)
```
```{r}
B <- ScaleData(B, split.by = "sample")
```

```{r}
# Find markers for dark zone and light zone germinal center B cells
Idents(B) <- B$CellTypes
DZvsLZ <- FindMarkers(B, ident.1 = "DZ", ident.2 = "LZ", features = features, slot = "scale.data")
head(DZvsLZ, n = 10)
```

```{r, fig.width = 8, fig.height = 5}
FeaturePlot(B, features = c("CCNB1", "AICDA", "FCRL2", "SLA"), split.by = "sample", pt.size=0.1, by.col = FALSE, slot = "data") +
  theme(legend.position = "right")
ggsave("Shared_Dominik/SDR005_data_share/20240205_SDR005_SDRranger_gDNA-V4_cDNA-V5/Supp7/Marker_UMAPs.pdf", width = 8, height = 5)
```

```{r, fig.width=7, fig.height=5}
# find markers for every State compared to all remaining cells, report only the positive
# ones
Idents(B) <- B$CellTypes
State_Markers <- FindAllMarkers(B, only.pos = TRUE)
State_Markers %>%
    group_by(cluster) %>%
    dplyr::filter(avg_log2FC > 1) %>%
    slice_head(n = 5) %>%
    ungroup() -> top5
DoHeatmap(B, features = top5$gene) + NoLegend()
```

## Add maturation state annotations to the full object

```{r}
# Step 1: Identify overlapping cells
overlapping_cells <- intersect(rownames(sobj@meta.data), rownames(Tumors_merged@meta.data))

# Ensure that 'cluster' or your actual annotation column name is a factor and update its levels
# Replace 'cluster' with the actual column name that holds your annotations
all_levels <- unique(c(as.character(sobj@meta.data$CellTypes), as.character(Tumors_merged@meta.data$predicted.State)))
sobj@meta.data$CellTypes <- factor(sobj@meta.data$CellTypes, levels = all_levels)


# Step 2: Extract annotations from Tumors for overlapping cells
# Replace 'cluster' with the actual column name that holds your annotations
new_annotations <- Tumors_merged@meta.data[overlapping_cells, "predicted.State"]

# Step 3: Update annotations in sobj for overlapping cells
sobj@meta.data[overlapping_cells, "CellTypes"] <- new_annotations

# Optional: Check some of the updated annotations
head(sobj@meta.data[overlapping_cells, ])
```


```{r}
sobj$CellTypes <- factor(sobj$CellTypes, 
                      levels = c("NaÃ¯ve", "DZ", "LZ", "Mem IgM", "Mem IgG", "Plasma", "Other"))
DimPlot(sobj, group.by = "CellTypes", cols = mathue10, shuffle = TRUE) +
  coord_fixed()

DimPlot(sobj, group.by = "CellTypes", reduction = "umap", label = FALSE, pt.size = 0.1, shuffle = TRUE, cols = mathue10) + 
  coord_fixed() + 
  labs(x = "UMAP 1", y = "UMAP 2", title = "Maturation States") +
  theme(legend.position = "bottom")
ggsave("Shared_Dominik/SDR005_data_share/20240205_SDR005_SDRranger_gDNA-V4_cDNA-V5/Supp7/MaturationUMAP.pdf", width = 4, height = 5)
```

```{r}
# Label all states except DZ and LZ as "B other"
Idents(sobj) <- "CellTypes"
new.cluster.ids <- c("B unassigned", "DZ", "LZ", "B unassigned", "B unassigned", "B unassigned", "Other")
names(new.cluster.ids) <- levels(sobj)
sobj <- RenameIdents(sobj, new.cluster.ids)
sobj$CellTypes2 <- Idents(sobj)

sobj$CellTypes2 <- factor(sobj$CellTypes2, 
                      levels = c("DZ", "LZ", "B unassigned", "Other"))
DimPlot(sobj, group.by = "CellTypes2")
```

```{r}
# Do the same for the column with unassigned cells
# Step 1: Identify overlapping cells
overlapping_cells <- intersect(rownames(sobj@meta.data), rownames(Tumors_merged@meta.data))

# Ensure that 'cluster' or your actual annotation column name is a factor and update its levels
# Replace 'cluster' with the actual column name that holds your annotations
all_levels <- unique(c(as.character(sobj@meta.data$CellTypes), as.character(Tumors_merged@meta.data$predicted.State3)))
sobj@meta.data$CellTypes3 <- factor(sobj@meta.data$CellTypes, levels = all_levels)


# Step 2: Extract annotations from Tumors for overlapping cells
# Replace 'cluster' with the actual column name that holds your annotations
new_annotations <- Tumors_merged@meta.data[overlapping_cells, "predicted.State3"]

# Step 3: Update annotations in sobj for overlapping cells
sobj@meta.data[overlapping_cells, "CellTypes3"] <- new_annotations

# Optional: Check some of the updated annotations
head(sobj@meta.data[overlapping_cells, ])

sobj$CellTypes3 <- factor(sobj$CellTypes3, 
                      levels = c("NaÃ¯ve", "DZ", "LZ", "Mem IgM", "Mem IgG", "Plasma", "B unassigned", "Other"))
DimPlot(sobj, group.by = "CellTypes3")
```

```{r, eval=FALSE}
saveRDS(sobj, "/g/huber/users/fitzgerald/R/scDNA-RNA/Shared_Dominik/SDR005_data_share/20240205_SDR005_SDRranger_gDNA-V4_cDNA-V5/SDR005_Combined_Geno_full.rds")
```



