---
title: "SDR002_Target_selection"
author: "Dominik Lindenhofer"
date: "2023-04-13"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


## Install/load libraries


```{r}
BiocManager::install("TAPseq")
BiocManager::install("BSgenome.Hsapiens.UCSC.hg38")
BiocManager::install("primer3_core")
BiocManager::install("makeblastdb")
BiocManager::install("blastn")

#Call primer3 and blast installation paths
Sys.setenv(PATH = paste("/Users/dominik.lindenhofer/opt/anaconda3/bin", Sys.getenv("PATH"), sep = ":"))
Sys.setenv(PATH = paste("/Users/dominik.lindenhofer/ncbi-blast-2.13.0+/bin", Sys.getenv("PATH"), 
                        sep = ":"))

library(TAPseq)
library(GenomicRanges)
library(BiocParallel)
library(BSgenome)
library(dplyr)



```




#Run TAP-seq pipeline - modify to only have one inner primer predicted

```{r}


# human genome BSgenome object (needs to be istalled from Bioconductor)
hg38 <- getBSgenome("BSgenome.Hsapiens.UCSC.hg38")

# download and import gencode hg38 annotations
url <- "ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_40/gencode.v40.annotation.gtf.gz"
annot <- import(url, format = "gtf")

# extract exon annotations for protein-coding genes to build transcripts
tx_exons <- annot[annot$type == "exon" & annot$gene_type == "protein_coding"]


# Names of target genes
target_gene_names <- read.delim("Input/RNA_seq_sampled.tsv") %>% dplyr::select(hgnc_symbol) %>% pull(1)

target_genes <- tx_exons[tx_exons$gene_name %in% target_gene_names]
target_genes <- split(target_genes, f = target_genes$gene_name)

print('target genes:')
print(target_genes)



# 10X scRNAseq read data 
#Public dataset was used from DOI 10.1101/gr.223925.117

scRNAseq_bam <- "possorted_genome_bam.bam"

# register backend for parallelization
register(MulticoreParam(workers = 10))

print('inferring polyA sites...')

# infer polyA sites from Drop-seq data
polyA_sites <- inferPolyASites(target_genes, bam = scRNAseq_bam, polyA_downstream = 50,
                               wdsize = 100, min_cvrg = 1, parallel = TRUE)



# In a real-world application it's adviced to
# manually inspect the polyA site predictions and remove any obvious false positives. This is easiest
# done by exporting the polyA sites into a .bed. This file can then be loaded into a genome browser
# (e.g. IGV) together with the target gene annotations and the .bam file to visually inspect the polyA
# site predictions.


export(polyA_sites, con = "./ipsc_polyA_sites.bed", format = "bed")
export(unlist(target_genes), con = "./ipsc_target_genes.gtf", format = "gtf")


# Once we are happy with our polyA predictions, we then use them to truncate the target gene
# transcripts. All target gene transcripts overlapping any polyA sites are truncated at the polyA
# sites, as we assume that they mark the transcripts 3' ends. By default the most downstream polyA
# site is taken to truncate a transcript. If multiple transcripts per gene overlap polyA sites, a
# merged truncated transcript model is generated from all overlapping transcripts. If no transcript
# of a given gene overlap with any polyA sites a consensus transcript model is created from all
# transcripts for that gene.

print('truncating transcripts...')

# truncate transcripts at inferred polyA sites
truncated_txs <- truncateTxsPolyA(target_genes, polyA_sites = polyA_sites, parallel = TRUE)

# To finalize this part, we need to extract the sequences of the truncated transcripts. Here we use
# Bioconductor's BSgenome to export transcript sequences, but the genome sequence could also be
# imported from a fasta file into a DNAStringSet object.


# get sequence for all truncated transcripts
txs_seqs <- getTxsSeq(truncated_txs, genome = hg38)


# Design  primers for MB

# MODIFICATION - only one primer predicted as there is no nested PCR done

# Primer3 uses boulder-IO records for input and output (see: http://primer3.org/manual.html). TAPseq
# implements TsIO and TsIOList objects, which store Primer3 input and output for TAP-seq primer
# design in R's S4 class system. They serve as the users interface to Primer3 during primer design.

# First we create Primer3 input for outer forward primer design based on the obtained transcript
# sequences. The reverse primer used in all PCR reactions is used, so that Primer3 can pick forward
# primers suitable to work with it. By default this function uses the 10x Beads-oligo-dT and right
# primer, and chooses optimal, minimum and maximum melting temperatures based on that. If another
# protocol (e.g. Drop-seq) is used, these parameters can be adapted.

print('creating inner primers...')


# create new TsIO objects for inner primers, note the different product size
inner_primers <- TAPseqInput(txs_seqs, target_annot = truncated_txs,
                             product_size_range = c(150, 300), primer_num_return = 5,
                             primer_opt_tm = 60,
                             primer_min_tm = 58,
                             primer_max_tm = 62)

# design inner primers
inner_primers <- designPrimers(inner_primers)

#Blast primers
print('creating BLASTDb')

# create blast database
blastdb <- file.path(tempdir(), "blastdb")
createBLASTDb(genome = hg38, annot = tx_exons, blastdb = blastdb)

help(createBLASTDb)
print('BLASTING inner primers')



print('BLASTing inner primers')

# blast inner primers
inner_primers <- blastPrimers(inner_primers, blastdb = blastdb, max_mismatch = 0,
                              min_aligned = 0.75)

print('picking best targets')
                              
# pick best primer per target gene
best_inner_primers <- pickPrimers(inner_primers, n = 1, by = "off_targets")


#Add MB primer for gDNA


#Load data
library(stringr)

MB_primers_check <- read.csv(file = "Input/MB_primers_for_checking_comp.csv")

#MB_primers_check <- MB_primers_check[c(1:5),]
MB_primers_check <- MB_primers_check %>% dplyr::mutate(index = row_number())
MB_primers_check <- MB_primers_check %>% dplyr::mutate(ID = str_c(AmpID, index, sep = "_" ))

#Make a loop to add the primers

#Modify with proper IRanges object classification

List_TsIO <- list()

for (i in unique(MB_primers_check$ID)){
  df_filtered <- MB_primers_check %>% dplyr::filter(ID == i)
  Template_primer <- IRanges(start = 1, width = str_length(df_filtered$Primer_wOverhangs), names = df_filtered$ID)
  Template_primer@elementMetadata <- DataFrame(list(sequence =  df_filtered$Primer_wOverhangs))
TsIO_i <- new("TsIO", sequence_id = df_filtered$ID)
TsIO_i@tapseq_primers <- Template_primer
List_TsIO <- append(List_TsIO, TsIO_i)
}




#Add those primers to the best inner primers

best_inner_primers_list <- best_inner_primers@listData
best_inner_primers_list_add <- c(best_inner_primers_list, List_TsIO)

TsIOList_TAP_MB <- TsIOList(best_inner_primers_list_add)


#Check for compatibility 

TsIOList_TAP_MB_comp <- checkPrimers(TsIOList_TAP_MB)


#Make dataframe to visualize
comp <- bind_rows(inner = TsIOList_TAP_MB_comp, .id = "set")


# add variable for pairs with any complemetarity score higher than 47
comp <- comp %>%
  mutate(high_compl = if_else(primer_pair_compl_any_th > 47 | primer_pair_compl_end_th > 47,
                              true = "high", false = "ok")) %>% 
  mutate(high_compl = factor(high_compl, levels = c("ok", "high")))

print('complementarity scores:')
print(comp)
                              
# plot complementarity scores
p <- ggplot(comp, aes(x = primer_pair_compl_any_th, y = primer_pair_compl_end_th)) +
  #facet_wrap(~set, ncol = 2) +
  geom_point(aes(color = high_compl), alpha = 0.25) +
  scale_color_manual(values = c("black", "red"), drop = FALSE) +
  geom_hline(aes(yintercept = 47), colour = "darkgray", linetype = "dashed") +
  geom_vline(aes(xintercept = 47), colour = "darkgray", linetype = "dashed") +
  labs(title = "Complementarity scores TAP-seq/MB primer combinations",
       color = "Complementarity") +
  theme_bw()


plot_save_dir <- "TAP_seq_multiplexed_PCR_testing"

ggsave(filename = "SDR002_inner_primer_MB_primer_comp_predictions.pdf", path = plot_save_dir, plot = p, device = "pdf", scale = 1,
       width = 15, height = 15, units = "cm", dpi = 300)



# No primer pairs with complementarity scores above 47 were found, so they should be ok to use in
# multiplex PCRs.

# Export primers

# To finish the workflow, we can export the designed primers to data.frames, which can be written to
# .csv files for easy storage of primer sets.

# create data.frames for outer and inner primers
inner_primers_df_add <- primerDataFrame(best_inner_primers)

write.csv(inner_primers_df,'ipsc_best_inner_primers.csv')


# Furthermore we can create BED tracks with the genomic coordinates of the primer binding sites for
# viewing in a genome browser.

# create BED tracks for outer and inner primers with custom colors
inner_primers_track <- createPrimerTrack(best_inner_primers, color = "goldenrod1")

# the output data.frames contain lines in BED format for every primer
head(inner_primers_track)

# export tracks to .bed files ("" writes to the standard output, replace by a file name)
exportPrimerTrack(inner_primers_track, con = "./ipsc_inner_primers.bed")


# Session information

#All of the output in this vignette was produced under the following conditions:
sessionInfo()
```



