---
title: "20230829_SDR001_Comparison"
author: "Dominik Lindenhofer"
date: "2023-08-29"
output: html_document
---

```{r}

```

##  Load libraries


```{r}

library(Seurat)
library(dplyr)
library(ggplot2)
library(Matrix)
library(reshape2)
library(data.table)
library(stringr)
library(tidyr)
library(scales)
library(rtracklayer)
library(data.table)
library(readr)


```

#Load both fully annotated Seurats

```{r}

cDNA_dCas9_data_Seurat <-readRDS("SDR001_dCas9_Geno_full.rds")
cDNA_PE_data_Seurat <-readRDS("SDR001_PE_Geno_full.rds")

#Merge the 2 seurats

cDNA_data_Seurat <-merge(cDNA_dCas9_data_Seurat, cDNA_PE_data_Seurat, project = "SDR001")

metadata <- cDNA_data_Seurat@meta.data
metadata$cell_merge <- rownames(metadata)

cDNA_data_Seurat@assays$RNA <- JoinLayers(cDNA_data_Seurat)
cDNA_data_Seurat@assays$gDNA <- JoinLayers(cDNA_data_Seurat@assays$gDNA )


#Make new Seurat with combined data - merge somehow gives an issue

RNA_mat <- cDNA_data_Seurat@assays$RNA@assays$RNA$counts %>% as.data.frame()
all.equal(colnames(RNA_mat), metadata$cell_merge )
gDNA_mat <- cDNA_data_Seurat@assays$gDNA$counts
all.equal(colnames(gDNA_mat), metadata$cell_merge )


#Make Seurat
cDNA_data_Seurat <- CreateSeuratObject(RNA_mat, project = "SDR001")

#Add metadata
all.equal(colnames(cDNA_data_Seurat), rownames(metadata))
cDNA_data_Seurat <- AddMetaData(object = cDNA_data_Seurat, metadata = metadata)


#Add gDNA as an assay


#Make assay and add to Seurat
gDNA_assay <- CreateAssay5Object(counts = as.sparse(gDNA_mat))
cDNA_data_Seurat[["gDNA"]] <- gDNA_assay
Assays(cDNA_data_Seurat)


#Prepare cell names for renaming
cells_dCas9 <- metadata %>% filter(orig.ident == "SDR001_dCas9") %>% select(cell_merge) %>% pull(1)
cells_PE <- metadata %>% filter(orig.ident == "SDR001_PE") %>% select(cell_merge) %>% pull(1)



#Load all variants again 

#dCas9

all_var_full_dCas9 <- read_tsv("06_GATK_VEP/CRISPRi.vcf", skip = 60, col_names = TRUE)
#Fill in empty information properly 
all_var_full_dCas9 <- all_var_full_dCas9  %>% mutate(ID = ".", QUAL = ".", FILTER = ".", FORMAT = "GT:AD:DP:GQ:PL")

#Rename cells to have suffix like in merged seurat
to_add_dCas9 <- all_var_full_dCas9[,c(1:9)]
all_var_dCas9 <- all_var_full_dCas9[,c(10:length(all_var_full_dCas9))]
colnames(all_var_dCas9) <- cells_dCas9
all_var_full_dCas9 <- cbind(to_add_dCas9, all_var_dCas9)



#PE
all_var_full_PE <- read_tsv("/06_VEP/PE.vcf", skip = 60, col_names = TRUE)
#Fill in empty information properly 
all_var_full_PE <- all_var_full_PE  %>% mutate(ID = ".", QUAL = ".", FILTER = ".", FORMAT = "GT:AD:DP:GQ:PL")

#Rename cells to have suffix like in merged seurat
to_add_PE <- all_var_full_PE[,c(1:9)]
all_var_PE <- all_var_full_PE[,c(10:length(all_var_full_PE))]
colnames(all_var_PE) <- cells_PE
all_var_full_PE <- cbind(to_add_PE, all_var_PE)



#Merge the variant files

all_var_full <- full_join(all_var_full_dCas9, all_var_full_PE, by = c("#CHROM", "POS", "ID", "REF", "ALT", "QUAL", "FILTER", "INFO", "FORMAT"))

#REMARK - use skripts now from other analysis 

```



#Make metadata for genoytpes

```{r}

#Split for processing and adding

#for metadata
to_add <- all_var_full[,c(1:9)]

#For variants
all_var <- all_var_full[,c(10:length(all_var_full))]
all_var <- as.data.table(all_var)


#_________________________________________________________________________

#Make metadata table for Genotypes

#Split info column - only first row now extracted - multiple entries per gene 
to_add_info <- lapply(to_add$INFO, function(x) unlist(strsplit(x, split = ",-")))
to_add_info_split <- data.frame(do.call('rbind', strsplit(as.character(to_add_info),'|',fixed=TRUE)))
to_add_info_split_for_merge <- data.frame(Consequence = to_add_info_split$X2, 
                                          IMPACT=to_add_info_split$X3, 
                                          SYMBOL=to_add_info_split$X4, 
                                          Gene= to_add_info_split$X5, 
                                          Feature_type =to_add_info_split$X6, 
                                          BIOTYPE =to_add_info_split$X7, 
                                          SNP_ID = to_add_info_split$X18)

#Add to metadata
to_add <- cbind(to_add, to_add_info_split_for_merge)



Geno_meta <- to_add

#Unique identifier for each variant
#Modify Consequence string so its compatible with seurat later

Geno_meta <- Geno_meta %>% mutate(Consequence = str_replace_all(Consequence, "_", ":"))
Geno_meta <- Geno_meta %>% mutate(ID_Geno = str_c(SYMBOL, `#CHROM`, POS, REF, ALT, Consequence ,sep = "-"))


#Annotate AMP_ID to each varint

annot <- read.delim("REF_lists/SDR001_gDNA_targets_annotate.txt")


Geno_meta_gr <- Geno_meta %>% select(Chromosome = `#CHROM`, Start = POS, END = POS, ID = ID_Geno)
Geno_meta_gr <-  makeGRangesFromDataFrame(Geno_meta_gr, keep.extra.columns = TRUE)

annot_gr <- annot %>% select(Chromosome, Start, End, ID = Amp_ID)
annot_gr <- makeGRangesFromDataFrame(annot_gr, keep.extra.columns = TRUE)



Overlap <- findOverlaps(Geno_meta_gr, annot_gr)
#gene_names <- CharacterList(split(SDR005_hg38_MB_input_gr$Gene[subjectHits(Overlap)],queryHits(Overlap)))
Gene<- split(annot_gr$ID[subjectHits(Overlap)], queryHits(Overlap))
Gene <- CharacterList(split(annot_gr$ID[subjectHits(Overlap)], queryHits(Overlap)))

mcols(Geno_meta_gr) <- DataFrame(mcols(Geno_meta_gr), Gene) 

#Make dataframe
Geno_meta_add <- Geno_meta_gr %>% as.data.frame() %>% tibble::as_tibble() 
Geno_meta_add  <- unnest(Geno_meta_add, cols = Gene) %>% distinct() %>% select(ID, Amp_ID = Gene)
Geno_meta <- Geno_meta %>% left_join(Geno_meta_add, by = c("ID_Geno" = "ID"))

#Also add in target name
annot_add <- annot %>% select(ID = Amp_ID, SNP_ID_annot = SNP_ID, Gene_annot = Gene)
Geno_meta <- Geno_meta %>% left_join(annot_add, by = c("Amp_ID" = "ID"))

#Unique identifier for each variant - modify with target_name from Julia and not with the one from VEP
#Modify Consequence string so its compatible with seurat later


Geno_meta <- Geno_meta %>% mutate(ID_Geno = str_c(Gene_annot, `#CHROM`, POS, REF, ALT, Consequence ,sep = "-"))



```




## Add full genotype data as an assay 
0 (WT), 1 (HET), 2 (HOM) - rest of information as list in @misc slot
No Cutoff first - do that in a second assay

```{r}

#For variants
all_var <- all_var_full[,c(10:length(all_var_full))]
all_var <- as.data.table(all_var)


#_________________________________________________________________________
#Genotype

#Extract the genotype information from the filtered variant file
#Get genotype
extract_vcf_info <- function(x) {
  sapply(strsplit(x, ":"), function(y) y[1])
}

all_var_GT <- all_var[, lapply(.SD, extract_vcf_info)]


#Replace 0/1 with 1 , 1/1 with 2, NA with 0

all_var_GT  <- all_var_GT %>%
  mutate(across(everything(), ~ case_when(
    . == "0/1" ~ "1",
    . == "1/1" ~ "2",
    TRUE ~ as.character(.)
  ))) %>%
  mutate(across(everything(), as.numeric)) %>%
  mutate(across(everything(), ~ replace_na(., 0)))


#Integreate 0/0 based on read coverage per cell 

cut_off = 0 # no cutoff at this point

gDNA_mat <- cDNA_data_Seurat@assays$gDNA$counts %>% as.data.frame()
all.equal(colnames(gDNA_mat), colnames(all_var))

gDNA_mat <- gDNA_mat %>% mutate(Amp_ID = rownames(gDNA_mat))

#Get outline of Geno - information 

gDNA_mat_mod <- data.frame(Amp_ID = Geno_meta$Amp_ID)

gDNA_mat_mod <- gDNA_mat_mod %>% left_join(gDNA_mat, by = c("Amp_ID"))
gDNA_mat_mod <- gDNA_mat_mod %>% select(-Amp_ID)
rownames(gDNA_mat_mod) <- Geno_meta$ID_Geno


gDNA_mat_mod_rep <- gDNA_mat_mod %>%   
    mutate(across(everything(), ~ case_when(
    . > cut_off ~ 0,
    . <= cut_off ~ NA
  )))


#Combine both matrices for a full Genotype matrix - NAs introduced if not covered by cutoff
all_var_GT <- as.matrix(all_var_GT) + as.matrix(gDNA_mat_mod_rep) %>% as.data.frame()
rownames(all_var_GT ) <- Geno_meta$ID_Geno

#Check order of cells again
all.equal(colnames(all_var_GT), colnames(cDNA_data_Seurat))




#_________________________________________________________________________
#Read depth 

#RD different in gDNA matrix and variant calling file 

#Fore read depth
extract_vcf_info <- function(x) {
  sapply(strsplit(x, ":"), function(y) y[3])
}

all_var_DP <- all_var[, lapply(.SD, extract_vcf_info)]
#all_var_DP <- cbind(to_add, all_var_DP)

all_var_DP <- all_var_DP %>% 
               mutate_all(as.numeric)

#Make DF for addition to produce NA values in gDNA mat for values that need to be substituted with read depth from GATK vcf file

all_var_DP_0 <- all_var_DP %>%
  mutate_all(~ifelse(is.na(.), 0, .)) %>% #if NA then to 0
  mutate_all(~ifelse(. > 0, NA, .)) %>% #if not 0 then NA
  mutate_all(as.numeric)

#Add this matrix with gDNA_mat_mod to produce NA values where read depth of DP should be - then add 0 there
gDNA_mat_mod <- as.matrix(gDNA_mat_mod) + as.matrix(all_var_DP_0) %>% as.data.frame()

#Make 0 in both for next addition to get rid of NA values
gDNA_mat_mod <- gDNA_mat_mod %>%  mutate_all(~ifelse(is.na(.), 0, .))
all_var_DP <- all_var_DP %>% mutate_all(~ifelse(is.na(.), 0, .)) 

#Now add the DP matrix with it to retain the original DP read depth 
all_var_DP <- as.matrix(gDNA_mat_mod) + as.matrix(all_var_DP) %>% as.data.frame()



#_________________________________________________________________________
#Calculate VAF with alternative reads

#For allelic reads
extract_vcf_info <- function(x) {
  sapply(strsplit(x, ":"), function(y) y[2])
}

all_var_AD <- all_var[, lapply(.SD, extract_vcf_info)]

ALT_reads <-  all_var_AD %>%
         mutate_all(~sub(".*,", "", .)) %>%
  mutate_all(as.numeric) %>% 
  mutate_all(~ifelse(is.na(.), 0, .)) 


VAF <- as.matrix(ALT_reads)/as.matrix(all_var_DP) %>% as.data.frame()  %>%  mutate_all(as.numeric) 


#_________________________________________________________________________
#Genotype Quality 


#Fore Genotype Quality 
extract_vcf_info <- function(x) {
  sapply(strsplit(x, ":"), function(y) y[4])
}

all_var_GQ <- all_var[, lapply(.SD, extract_vcf_info)]
#all_var_GQ <- cbind(to_add, all_var_GQ)

all_var_GQ <- all_var_GQ %>% 
               mutate_all(as.numeric) 


#_________________________________________________________________________
#Add to Seurat 

#GT as assay
#Geno_meta as metadata
#rest of information (all_var, RD, VAF, GQ) as list

all.equal(rownames(all_var_GT), Geno_meta$ID_Geno)
all.equal(colnames(all_var_GT), colnames(cDNA_data_Seurat))


#Make assay and add to Seurat
GT_assay <- CreateAssay5Object(counts = as.sparse(all_var_GT))
cDNA_data_Seurat[["Geno"]] <- GT_assay
Assays(cDNA_data_Seurat)

#Add metadata
cDNA_data_Seurat@assays$Geno@meta.data <- Geno_meta


#Make list for the rest of the information 
all_var <- all_var %>% as.data.frame()
rownames(all_var) <- Geno_meta$ID_Geno

all_var_DP <- all_var_DP %>% as.data.frame()
rownames(all_var_DP) <- Geno_meta$ID_Geno

VAF <- VAF %>% as.data.frame()
rownames(VAF) <- Geno_meta$ID_Geno



all_var_GQ <- all_var_GQ %>% as.data.frame()
rownames(all_var_GQ) <- Geno_meta$ID_Geno

list_to_add <- list(full = all_var,
                    DP = all_var_DP,
                    VAF = VAF,
                    GQ = all_var_GQ)

#Add list to misc slot
cDNA_data_Seurat@assays$Geno@misc <- list_to_add


#Save again
saveRDS(cDNA_data_Seurat, "SDR001_Combined_Geno_full.rds")


```



#Apply filtering on read depth and GQ

```{r}

#Load data

cDNA_data_Seurat <- readRDS("SDR001_Combined_Geno_full.rds")

#Get all matrices
GT_mat <- cDNA_data_Seurat@assays$Geno$counts %>% as.data.frame()
Geno_meta <- cDNA_data_Seurat@assays$Geno@meta.data

Geno_list <- cDNA_data_Seurat@assays$Geno@misc
list2env(Geno_list, envir = .GlobalEnv)


#Set cutoffs 

cut_off_RD = 5
cut_off_GQ = 30

#Make RD cutoff - make data frame dataframe to produce NAs 
DP_rep <- DP %>%   
    mutate(across(everything(), ~ case_when(
    . > cut_off_RD ~ 0,
    . <= cut_off_RD ~ NA
  )))

#Make GQ cutoff - make data frame dataframe to produce NAs 

GQ_rep <- GQ %>%
  mutate(across(everything(), ~ case_when(
    is.na(.) ~ 99,            #Set NAs to maximum value - no WT data integrated here
    TRUE ~ as.numeric(.)
  ))) %>%
  mutate(across(everything(), ~ case_when(
    . > cut_off_GQ ~ 0,
    . <= cut_off_GQ ~ NA
  )))


#Add in all matrixes with GT to produce NAs 

GT_mat_mod <- as.matrix(GT_mat) + as.matrix(DP_rep) + as.matrix(GQ_rep) %>% as.data.frame()


#Add back into new assay


#Make assay and add to Seurat
GT_mod_assay <- CreateAssay5Object(counts = as.sparse(GT_mat_mod))
cDNA_data_Seurat[["Geno_filtered"]] <- GT_mod_assay
Assays(cDNA_data_Seurat)

#Add metadata
cDNA_data_Seurat@assays$Geno_filtered@meta.data <- Geno_meta

#Add list data - unfiltered

cDNA_data_Seurat@assays$Geno_filtered@misc <- Geno_list

#Save again
saveRDS(cDNA_data_Seurat, "SDR001_Combined_Geno_full.rds")

```


#Visualize nCells per Variant


```{r}

Geno_mat <- cDNA_data_Seurat@assays$Geno_filtered$counts %>% as.data.frame()
Geno_mat_mod <- Geno_mat %>% mutate_all(~na_if(., 0))
nCells_per_Variant <- rowSums(!is.na(Geno_mat_mod ))
nCells_per_Variant <- data.frame(nCells_per_Variant = nCells_per_Variant)

#Quickly visualize

ggplot(nCells_per_Variant, aes(x=nCells_per_Variant)) + 
  scale_x_log10()+
  geom_histogram(binwidth=0.1)

```


#Filter for variants that are intended to be targeted

```{r}

variants_annot <- read.csv("REF_lists/SDR001_variants_annotate.csv")

Geno_meta <- cbind(Geno_meta, nCells_per_Variant)

Geno_meta_filter <- Geno_meta %>% left_join(variants_annot, by = "Amp_ID")
Geno_meta_filter <- Geno_meta_filter %>% filter(POS == SNP_pos_hg38)




```


#Plot some quality control plots - VAF vs read depth - add VAF as an assay 

```{r}


#Load data

cDNA_data_Seurat <- readRDS("SDR001_Combined_Geno_full.rds")

metadata <- cDNA_data_Seurat@meta.data

#Get all matrices
GT_mat <- cDNA_data_Seurat@assays$Geno_filtered$counts %>% as.data.frame()
Geno_meta <- cDNA_data_Seurat@assays$Geno_filtered@meta.data

Geno_list <- cDNA_data_Seurat@assays$Geno_filtered@misc
list2env(Geno_list, envir = .GlobalEnv)




#_________________________________________________________________________
#Respahe data for plotting

#Genotype - still need to add in WT coverage !!! - done somewhat above already
GT_mat <- GT_mat %>% t() %>% as.data.frame()
GT_mat <- GT_mat %>% mutate(cell = rownames(GT_mat))
GT_mat_res <- GT_mat %>% reshape2::melt(id.vars = c("cell"),
                                        variable.name = "ID", 
                                        value.name = "Genotype")


#VAF
VAF <- VAF %>% t() %>% as.data.frame()
VAF <- VAF %>% mutate(cell = rownames(VAF))
VAF_res <- VAF %>% reshape2::melt(id.vars = c("cell"),
                                  variable.name = "ID", 
                                  value.name = "VAF")



#Read depth 
DP <- DP %>% t() %>% as.data.frame()
DP <- DP %>% mutate(cell = rownames(DP))
DP_res <- DP %>% reshape2::melt(id.vars = c("cell"),
                                variable.name = "ID", 
                                value.name = "DP")




#Genotype quality 
GQ <- GQ  %>% t() %>% as.data.frame()
GQ  <- GQ  %>% mutate(cell = rownames(GQ))
GQ_res <- GQ  %>% reshape2::melt(id.vars = c("cell"),
                                variable.name = "ID", 
                                value.name = "GQ")



#Join in all of them 

Combined_VAF_to_plot <- GT_mat_res %>% left_join(VAF_res, by = c("cell" = "cell", "ID" = "ID"))
Combined_VAF_to_plot <- Combined_VAF_to_plot %>% left_join(DP_res, by = c("cell" = "cell", "ID" = "ID"))
Combined_VAF_to_plot <- Combined_VAF_to_plot %>% left_join(GQ_res, by = c("cell" = "cell", "ID" = "ID"))

#Add condition info

to_add <- metadata %>% select(cell_merge, Type)
info_split <- data.frame(do.call('rbind', strsplit(as.character(to_add$Type),'_',fixed=TRUE)))
colnames(info_split) <- c("Type2", "sbc_number")
to_add <- to_add %>% mutate(Type2 = info_split$Type2)
to_add  <- to_add  %>%
    mutate(Type3 = case_when(Type == "PE_1" | Type == "PE_2" ~ "PEmax-MLHdn1",
                               Type == "PE_3" | Type == "PE_4" ~ "PEmax",
                               Type2 == "dCas9" ~ "dCas9"))



Combined_VAF_to_plot <- Combined_VAF_to_plot %>% left_join(to_add, by = c("cell" = "cell_merge"))

unique(Combined_VAF_to_plot$Genotype)
#Filter out NA values - change later 

#Combined_VAF_to_plot <- Combined_VAF_to_plot %>% filter(!is.na(Genotype))
Combined_VAF_to_plot$Genotype <- factor(Combined_VAF_to_plot$Genotype, levels = c(2, 1, 0, NA))


#_________________________________________________________________________
#Plot VAF vs read depth - Color genotype

#Only do plots for variants that we wanted to find

out_put_dir <- "SDR001_Comb_GATK/All_variants_VAF_plots_intended"
plot_genes <- unique(Geno_meta_filter$ID_Geno)



i=plot_genes[28]

#group_colors <- c("1/1" = "forestgreen", "0/1" = "cyan2", "0/0" = "dodgerblue3", "./." = "grey")
group_colors <- c("2" = "forestgreen", "1" = "cyan2", "0" = "dodgerblue3", "NA" = "grey")


for (i in unique(plot_genes)){
  Df_to_plot_i <- Combined_VAF_to_plot %>% filter(ID == i)
  
  p1 <-   Df_to_plot_i %>%
    ggplot(aes(x=VAF, y=DP, color=Genotype)) + 
  	geom_point(size = 5) + 
	#scale_colour_gradient(low = "gray90", high = "#990000", limits = c(0,1)) +
  #xlim(c(0,30))+
    #ylim(c(0,30))+
    scale_color_manual(values = group_colors )+
  	scale_x_continuous(limits = c(0,1)) + 
  	scale_y_log10(limits = c(1,10000)) + 
  	theme_classic() +
  	#geom_vline(xintercept = 10) +
  	#geom_hline(yintercept = 10) +
    	theme(plot.title = element_text(hjust=0.5, face="bold")) +
    facet_wrap("Type2")+
  	ggtitle(i)
  
  
  p2 <- Df_to_plot_i %>%
    ggplot(aes(x=VAF, y=DP, color=GQ)) + 
  	geom_point(size = 5) + 
	scale_colour_gradient(low = "gray90", high = "#990000", limits = c(1,99)) +
  #xlim(c(0,30))+
    #ylim(c(0,30))+
    #scale_color_manual(values = group_colors )+
  	scale_x_continuous(limits = c(0,1)) + 
  	scale_y_log10(limits = c(1,10000)) + 
  	theme_classic() +
  	#geom_vline(xintercept = 10) +
  	#geom_hline(yintercept = 10) +
    	theme(plot.title = element_text(hjust=0.5, face="bold")) +
        facet_wrap("Type2")+
  	ggtitle(i)
  
p <- p1 + p2
  

  ggsave(paste(i,".pdf",sep = ""), path = out_put_dir, plot = p, device = "pdf", scale = 1,
         width = 60, height = 15, units = "cm", dpi = 300)
    
}


p1
p


```



#Plot editing efficiency




```{r}



#Filter for variants that were intended to make

variants_annot <- read.csv("REF_lists/SDR001_variants_annotate.csv")
variants_annot <- variants_annot %>% mutate(chrom = str_c("chr", chrom, sep = ""))

variants_annot <- variants_annot %>% mutate(ID_variant = str_c(Gene, chrom, SNP_pos_hg38, REF, ALT, sep = "-"))


Geno_meta <- cDNA_data_Seurat@assays$Geno_filtered@meta.data
Geno_meta <- Geno_meta %>% mutate(ID_variant = str_c(Gene_annot, `#CHROM`, POS, REF, ALT, sep = "-"))


to_add <- Geno_meta %>% select(ID_Geno, ID_variant)

Combined_VAF_to_plot <- Combined_VAF_to_plot %>% left_join(to_add, by = c("ID" = "ID_Geno"))



#Filter for variants to look for 

to_filter <- variants_annot$ID_variant
Combined_VAF_to_plot_filter <- Combined_VAF_to_plot %>% filter(ID_variant %in% to_filter)



 
nCells <- Combined_VAF_to_plot %>%
  group_by(Type3) %>%
  summarise(nCells = n_distinct(cell))

nCells_dCas9 <- nCells %>% filter(Type3 == "dCas9") %>% pull(2)
nCells_PEmax <- nCells %>% filter(Type3 == "PEmax") %>% pull(2)
nCells_PEmax_MLHdn1 <- nCells %>% filter(Type3 == "PEmax-MLHdn1") %>% pull(2)

#dCas9

dCas9_HDR <- Combined_VAF_to_plot_filter %>% 
  filter(Type3 == "dCas9") %>%
 dplyr::filter(Genotype == "2" | Genotype == "1") %>%
  group_by(ID_variant) %>% 
  dplyr::summarise(count = n()) %>%
mutate(Type = "dCas9", nCells = nCells_dCas9) %>%
  mutate(edit_perc = count/nCells*100) 

#No cells found for these targets with HET or ALT genotype - make manual file for merging
to_add <- data.frame(ID_variant = setdiff(variants_annot$ID_variant, dCas9_HDR$ID_variant), count = 0, Type = "dCas9", nCells = 0, edit_perc = 0)


dCas9_HDR <- rbind(dCas9_HDR, to_add)


#PEmax

PEmax_HDR <- Combined_VAF_to_plot_filter %>% 
  filter(Type3 == "PEmax") %>%
 dplyr::filter(Genotype == "2" | Genotype == "1") %>%
  group_by(ID_variant) %>% 
  dplyr::summarise(count = n()) %>%
mutate(Type = "PEmax", nCells = nCells_PEmax) %>%
  mutate(edit_perc = count/nCells*100) 

#No cells found for these targets with HET or ALT genotype - make manual file for merging
to_add <- data.frame(ID_variant = setdiff(variants_annot$ID_variant, PEmax_HDR$ID_variant), count = 0, Type = "PEmax", nCells = 0, edit_perc = 0)


PEmax_HDR <- rbind(PEmax_HDR, to_add)



#PEmax_MLHdn1

PEmax_MLHdn1_HDR <- Combined_VAF_to_plot_filter %>% 
  filter(Type3 == "PEmax-MLHdn1") %>%
 dplyr::filter(Genotype == "2" | Genotype == "1") %>%
  group_by(ID_variant) %>% 
  dplyr::summarise(count = n()) %>%
mutate(Type = "PEmax-MLHdn1", nCells = nCells_PEmax_MLHdn1) %>%
  mutate(edit_perc = count/nCells*100) 

#No cells found for these targets with HET or ALT genotype - make manual file for merging
to_add <- data.frame(ID_variant = setdiff(variants_annot$ID_variant, PEmax_MLHdn1_HDR$ID_variant), count = 0, Type = "PEmax-MLHdn1", nCells = 0, edit_perc = 0)


PEmax_MLHdn1_HDR <- rbind(PEmax_MLHdn1_HDR, to_add)




#Add everything together

Edit_plot <- rbind(dCas9_HDR, PEmax_HDR, PEmax_MLHdn1_HDR)

#Remove sites that are het

to_filter <- dCas9_HDR %>% filter(edit_perc < 0.1) %>% select(ID_variant) %>% pull(1)

Edit_plot <- Edit_plot %>% filter(ID_variant %in% to_filter)

#PLOT

#Set order and colors
Edit_plot$Type <- factor(Edit_plot$Type, levels = c("dCas9", "PEmax", "PEmax-MLHdn1"))
group_colors <- c("dCas9" = "#0C2D57", "PEmax" = "#FC6736", "PEmax-MLH1dn" = "#FFB0B0" )



title = "dCas9_PE_summed_edits_perc.pdf"
p <- Edit_plot %>%
   ggplot(aes(x= Type, y = edit_perc))+ 
    stat_summary(aes(fill = Type), fun.y = mean, geom = "bar", alpha = 0.4, show.legend = FALSE) +
      geom_jitter(aes(color = Type),width = 0.25, size = 7, show.legend = FALSE)+
    stat_summary(fun.data = mean_se,geom = "errorbar", width = 0.25, size = 1)+
  scale_color_manual(values = group_colors)+
   scale_fill_manual(values = group_colors)+
      #coord_trans(y = "log10", limy = c(0,6000))+ 
      #scale_y_continuous(breaks = 10^(0:4)) +
  	theme_classic() +
  labs(y = "% of summed edits (HET and ALT)")+
  	theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  	theme(plot.title = element_text(hjust=0.5, face="bold")) +
  	ggtitle(title)

p


ggsave(filename = title, path = "SDR001_Comb_GATK", plot = p, device = "pdf", scale = 1,
       width = 10, height = 15, units = "cm", dpi = 300)



#Arrange according to editing efficiency 


Edit_plot <- Edit_plot %>% arrange(desc(edit_perc))
Edit_plot$ID_variant <- as.factor(Edit_plot$ID_variant) 
Edit_plot$Type <- as.character(Edit_plot$Type)
Edit_plot <- Edit_plot %>%
  arrange(desc(edit_perc)) %>%
  mutate(
    ID_variant = factor(ID_variant, levels = unique(ID_variant))  # Set levels in the order of appearance
  )

title = "dCas9_PE_summed_edits_perc_per_variant.pdf"
p <- Edit_plot %>%
  filter(Type != "dCas9") %>%
   ggplot(aes(x= ID_variant, y = edit_perc))+ 
    stat_summary(aes(fill = ID_variant), fun.y = mean, geom = "bar", alpha = 0.4, show.legend = FALSE) +
      geom_jitter(aes(color = Type),width = 0.25, size = 7, show.legend = FALSE)+
    stat_summary(fun.data = mean_se,geom = "errorbar", width = 0.25, size = 1)+
  scale_color_manual(values = group_colors)+
   scale_fill_manual(values = group_colors)+
      #coord_trans(y = "log10", limy = c(0,6000))+ 
      #scale_y_continuous(breaks = 10^(0:4)) +
  	theme_classic() +
  labs(y = "% of summed edits (HET and ALT)")+
  #scale_y_continuous(limits = c(0,5))+
  	theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  	theme(plot.title = element_text(hjust=0.5, face="bold")) +
  	ggtitle(title)

p

ggsave(filename = title, path = "SDR001_Comb_GATK", plot = p, device = "pdf", scale = 1,
       width = 30, height = 15, units = "cm", dpi = 300)








```







