---
title: "20231117_SDR005_get_ref_seqs_for_STAR"
author: "Dominik Lindenhofer"
date: "2023-11-17"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Load libs


```{r}


library(BSgenome)
library(Biostrings)
library(dplyr)

library(GenomicRanges)


```




```{r}

#Get seqs for V3 gDNA reference 


#get sequences for gDNA targets
SDR005_gDNA_amplicons <- read.delim("06_BCL.amplicons", header = FALSE)
colnames(SDR005_gDNA_amplicons) <- c("Chromosome", "Start", "End", "ID")


gDNA_targets_gr_1 <- makeGRangesFromDataFrame(SDR005_gDNA_amplicons, keep.extra.columns = TRUE)
gDNA_targets_gr_2 <- makeGRangesFromDataFrame(SDR005_gDNA_amplicons, keep.extra.columns = TRUE)



Overlap <- countOverlaps(gDNA_targets_gr_1, gDNA_targets_gr_2)
gDNA_targets_gr_1$count_overlaps <- Overlap
gDNA_targets_df_1 <- as.data.frame(gDNA_targets_gr_1)


```




```{r}
# human genome BSgenome object (needs to be istalled from Bioconductor)
hg38 <- getBSgenome("BSgenome.Hsapiens.UCSC.hg38")

#get sequences for gDNA targets
SDR005_gDNA_amplicons <- read.delim("06_BCL.amplicons", header = FALSE)
colnames(SDR005_gDNA_amplicons) <- c("Chromosome", "Start", "End", "ID")


gDNA_targets <- SDR005_gDNA_amplicons 


gDNA_targets_seq <- data.frame()

for (i in unique(gDNA_targets$ID)){
  gDNA_targets_i <- gDNA_targets %>% dplyr::filter(gDNA_targets$ID == i)
  gDNA_targets_dna_i <- as.character(Biostrings::getSeq(hg38, gDNA_targets_i$Chromosome, gDNA_targets_i$Start,	gDNA_targets_i$End))
  gDNA_targets_dna_i_df <- data.frame(ID = i, Chromosome = gDNA_targets_i$Chromosome, Start = gDNA_targets_i$Start, End = gDNA_targets_i$End,  Sequence = gDNA_targets_dna_i)
  gDNA_targets_seq <- rbind(gDNA_targets_seq, gDNA_targets_dna_i_df)
}


to_write <- gDNA_targets_seq %>% dplyr::select(ID, Sequence)

write.table(to_write, file = "06_BCL_gDNA_targets_seq.txt", quote = FALSE, row.names = FALSE, col.names = TRUE, sep = "\t")



```





#Make annotation file to be used for other seurat later

```{r}

#get sequences for gDNA targets
SDR005_gDNA_amplicons <- read.delim("06_BCL.amplicons", header = FALSE)
colnames(SDR005_gDNA_amplicons) <- c("Chromosome", "Start", "End", "ID")



#Add columns indicating positional information in amplicon 

SDR005_gDNA_amplicons  <- SDR005_gDNA_amplicons %>% mutate(Start_Amp = 1, End_Amp = End-Start+1 )




#Load file for annotation - from input for Mission Bio tapestri pipeline

SDR005_hg38_MB_input <- read.delim("06_BCL_hg38.bed", header = FALSE) %>% dplyr::select(Chromosome = V1, Start = V2, End = V3, Gene = V4)


#Find Overlaps and add info

SDR005_gDNA_amplicons_gr <- makeGRangesFromDataFrame(SDR005_gDNA_amplicons, keep.extra.columns = TRUE)
SDR005_hg38_MB_input_gr <- makeGRangesFromDataFrame(SDR005_hg38_MB_input, keep.extra.columns = TRUE)


Overlap <- findOverlaps(SDR005_gDNA_amplicons_gr, SDR005_hg38_MB_input_gr)
#gene_names <- CharacterList(split(SDR005_hg38_MB_input_gr$Gene[subjectHits(Overlap)],queryHits(Overlap)))
Gene<- split(SDR005_hg38_MB_input_gr$Gene[subjectHits(Overlap)], queryHits(Overlap))
Gene <- CharacterList(split(SDR005_hg38_MB_input_gr$Gene[subjectHits(Overlap)], queryHits(Overlap)))

mcols(SDR005_gDNA_amplicons_gr) <- DataFrame(mcols(SDR005_gDNA_amplicons_gr), Gene) 

#Write to dataframe
SDR005_gDNA_amplicons_ass <- SDR005_gDNA_amplicons_gr %>% as.data.frame() %>% tibble::as_tibble() 
SDR005_gDNA_amplicons_ass <- unnest(SDR005_gDNA_amplicons_ass, cols = Gene) %>% distinct()

write.table(SDR005_gDNA_amplicons_ass, file = "06_BCL_genomic_annotation_file.txt", quote = FALSE, row.names = FALSE, col.names = TRUE, sep = "\t")
```



