---
title: "20220804_ipsc_tapseq_primer_design_script_V1"
output: html_document
date: "2022-08-04"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Install/load libraries


```{r}
BiocManager::install("TAPseq")
BiocManager::install("BSgenome.Hsapiens.UCSC.hg38")
BiocManager::install("primer3_core")
BiocManager::install("makeblastdb")
BiocManager::install("blastn")

#Call primer3 and blast installation paths
Sys.setenv(PATH = paste("/Users/dominik.lindenhofer/opt/anaconda3/bin", Sys.getenv("PATH"), sep = ":"))
Sys.setenv(PATH = paste("/Users/dominik.lindenhofer/ncbi-blast-2.13.0+/bin", Sys.getenv("PATH"), 
                        sep = ":"))

library(TAPseq)
library(GenomicRanges)
library(BiocParallel)
library(BSgenome)
library(here)



```

## Run TAP-seq pipeline

```{r}

# human genome BSgenome object (needs to be istalled from Bioconductor)
hg38 <- getBSgenome("BSgenome.Hsapiens.UCSC.hg38")

# download and import gencode hg38 annotations
url <- "ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_40/gencode.v40.annotation.gtf.gz"
annot <- import(url, format = "gtf")

# extract exon annotations for protein-coding genes to build transcripts
tx_exons <- annot[annot$type == "exon" & annot$gene_type == "protein_coding"]

# Names of target genes
target_gene_names <- as.character(read.csv('Input/RNA_gene_list.csv',header = FALSE)[,1])

target_genes <- tx_exons[tx_exons$gene_name %in% target_gene_names]
target_genes <- split(target_genes, f = target_genes$gene_name)

print('target genes:')
print(target_genes)

# 10X scRNAseq read data 
#Public dataset was used from DOI 10.1101/gr.223925.117

scRNAseq_bam <- "possorted_genome_bam.bam"

# register backend for parallelization
register(MulticoreParam(workers = 4))

print('inferring polyA sites...')

# infer polyA sites from Drop-seq data
polyA_sites <- inferPolyASites(target_genes, bam = scRNAseq_bam, polyA_downstream = 50,
                               wdsize = 100, min_cvrg = 1, parallel = TRUE)

# In a real-world application it's adviced to
# manually inspect the polyA site predictions and remove any obvious false positives. This is easiest
# done by exporting the polyA sites into a .bed. This file can then be loaded into a genome browser
# (e.g. IGV) together with the target gene annotations and the .bam file to visually inspect the polyA
# site predictions.

library(rtracklayer)
export(polyA_sites, con = "./ipsc_polyA_sites.bed", format = "bed")
export(unlist(target_genes), con = "./ipsc_target_genes.gtf", format = "gtf")


# Once we are happy with our polyA predictions, we then use them to truncate the target gene
# transcripts. All target gene transcripts overlapping any polyA sites are truncated at the polyA
# sites, as we assume that they mark the transcripts 3' ends. By default the most downstream polyA
# site is taken to truncate a transcript. If multiple transcripts per gene overlap polyA sites, a
# merged truncated transcript model is generated from all overlapping transcripts. If no transcript
# of a given gene overlap with any polyA sites a consensus transcript model is created from all
# transcripts for that gene.

print('truncating transcripts...')

# truncate transcripts at inferred polyA sites
truncated_txs <- truncateTxsPolyA(target_genes, polyA_sites = polyA_sites, parallel = TRUE)

# To finalize this part, we need to extract the sequences of the truncated transcripts. Here we use
# Bioconductor's BSgenome to export transcript sequences, but the genome sequence could also be
# imported from a fasta file into a DNAStringSet object.

# human genome BSgenome object (needs to be istalled from Bioconductor)
hg38 <- getBSgenome("BSgenome.Hsapiens.UCSC.hg38")

# get sequence for all truncated transcripts
txs_seqs <- getTxsSeq(truncated_txs, genome = hg38)

# Design outer primers

# Primer3 uses boulder-IO records for input and output (see: http://primer3.org/manual.html). TAPseq
# implements TsIO and TsIOList objects, which store Primer3 input and output for TAP-seq primer
# design in R's S4 class system. They serve as the users interface to Primer3 during primer design.

# First we create Primer3 input for outer forward primer design based on the obtained transcript
# sequences. The reverse primer used in all PCR reactions is used, so that Primer3 can pick forward
# primers suitable to work with it. By default this function uses the 10x Beads-oligo-dT and right
# primer, and chooses optimal, minimum and maximum melting temperatures based on that. If another
# protocol (e.g. Drop-seq) is used, these parameters can be adapted.

print('creating outer primers...')

# create TAPseq IO for outer forward primers from truncated transcript sequences
outer_primers <- TAPseqInput(txs_seqs, target_annot = truncated_txs,
                             product_size_range = c(350, 500), primer_num_return = 5,
                             primer_opt_tm = 60,
                             primer_min_tm = 58,
                             primer_max_tm = 62
                             )


# We can now use Primer3 to design Primers and add them to the objects.
# design 5 outer primers for each target gene
outer_primers <- designPrimers(outer_primers)

# The output TsIO objects contain the designed primers and expected amplicons:

# get primers and pcr products for specific genes
# tapseq_primers(outer_primers$HBE1)
# pcr_products(outer_primers$HBE1)

# these can also be accessed for all genes
tapseq_primers(outer_primers)
pcr_products(outer_primers)

print('created primers:')

print(tapseq_primers(outer_primers))
print(pcr_products(outer_primers))

# We have now successfully designed 5 outer primers per target gene. To select the best primer per
# gene we could just pick the primer with the lowest penalty score. However, we will now use BLAST to
# try to estimate potential off-target priming for every primer. We will use this to then select the
# best primer per gene, i.e. the primer with the fewest off-targets.

# BLAST primers

# Creating a BLAST database containing all annotated transcripts and chromosome sequences takes a
# couple of minutes (and ~2Gb of free storage). For this example, we can generate a limited BLAST
# database, containing only transcripts of all target genes and the sequence of chromosome 11.

## chromosome 11 sequence
#chr11_genome <- DNAStringSet(getSeq(hg38, "chr11"))
#names(chr11_genome) <- "chr11"
## create blast database
#blastdb <- file.path(tempdir(), "blastdb")
#createBLASTDb(genome = chr11_genome, annot = unlist(target_genes), blastdb = blastdb)

# In a real-world application one would generate a database containing the entire genome and
# transcriptome. The database can be saved in a location and used for multiple primer designs whithout
# having to rebuild it everytime.

library(BSgenome)

# human genome BSgenome object (needs to be istalled from Bioconductor)
hg38 <- getBSgenome("BSgenome.Hsapiens.UCSC.hg38")

# download and import gencode hg38 annotations
url <- "ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_40/gencode.v40.annotation.gtf.gz"
annot <- import(url, format = "gtf")

# extract exon annotations for protein-coding genes to build transcripts
tx_exons <- annot[annot$type == "exon" & annot$gene_type == "protein_coding"]

print('creating BLASTDb')

# create blast database
blastdb <- file.path(tempdir(), "blastdb")
createBLASTDb(genome = hg38, annot = tx_exons, blastdb = blastdb)

print('BLASTING outer primers')

# Once we have our BLAST database, we can use it to estimate the number of exonic, intronic and
# intergenic off-targets for all primers. Note, that the number of exonic and intronic off-targets is
# provided on the level of genes, i.e. one exonic off-target means that a given primer might bind in
# exonic region(s) of one other gene. If the off-target binding site overlaps with exons of two genes,
# it will be counted twice, as it could bind to transcripts of two genes.
# now we can blast our outer primers against the created database
outer_primers <- blastPrimers(outer_primers, blastdb = blastdb, max_mismatch = 0,
                              min_aligned = 0.75)
                              
# the primers now contain the number of estimated off-targets
print('Estimates of off-targets:')
print(tapseq_primers(outer_primers))

# To finalize our set of outer primers, we want to choose the best primer per target gene, i.e. the
# one with the fewest exonic, intronic and intergenic off-target hits (in that order).

# select best primer per target gene based on the number of potential off-targets
best_outer_primers <- pickPrimers(outer_primers, n = 1, by = "off_targets")

# each object now only contains the best primer
print('after picking best primers')
print(tapseq_primers(best_outer_primers))

# Design inner primers

# To design nested inner primers we simply need to repeat the same procedure with a smaller product
# size range.

print('creating inner primers')

# create new TsIO objects for inner primers, note the different product size
inner_primers <- TAPseqInput(txs_seqs, target_annot = truncated_txs,
                             product_size_range = c(150, 300), primer_num_return = 5,
                             primer_opt_tm = 60,
                             primer_min_tm = 58,
                             primer_max_tm = 62)
                              
# design inner primers
inner_primers <- designPrimers(inner_primers)

print('BLASTing inner primers')

# blast inner primers
inner_primers <- blastPrimers(inner_primers, blastdb = blastdb, max_mismatch = 0,
                              min_aligned = 0.75)

print('picking best targets')
                              
# pick best primer per target gene
best_inner_primers <- pickPrimers(inner_primers, n = 1, by = "off_targets")

# Done! We succesfully designed TAP-seq outer and inner primer sets for our target gene panel.

# Multiplex compatibility

# As an additional step, we can verify if the designed primer sets are compatible for PCR
# multiplexing. For that we use Primer3's "check_primers" functionality:

# use checkPrimers to run Primer3's "check_primers" functionality for every possible primer pair
outer_comp <- checkPrimers(best_outer_primers)
inner_comp <- checkPrimers(best_inner_primers)

# We can now for instance plot the estimated complementarity scores for every pair. We highlight
# pairs with a score higher than 47, which is considered "critical" by Primer3 during primer design
# (see Primer3 for more information).

library(dplyr)
library(ggplot2)

# merge outer and inner complementarity scores into one data.frame
comp <- bind_rows(outer = outer_comp, inner = inner_comp, .id = "set")

# add variable for pairs with any complemetarity score higher than 47
comp <- comp %>%
  mutate(high_compl = if_else(primer_pair_compl_any_th > 47 | primer_pair_compl_end_th > 47,
                              true = "high", false = "ok")) %>% 
  mutate(high_compl = factor(high_compl, levels = c("ok", "high")))

print('complementarity scores:')
print(comp)
                              
# plot complementarity scores
#ggplot(comp, aes(x = primer_pair_compl_any_th, y = primer_pair_compl_end_th)) +
#  facet_wrap(~set, ncol = 2) +
#  geom_point(aes(color = high_compl), alpha = 0.25) +
#  scale_color_manual(values = c("black", "red"), drop = FALSE) +
#  geom_hline(aes(yintercept = 47), colour = "darkgray", linetype = "dashed") +
#  geom_vline(aes(xintercept = 47), colour = "darkgray", linetype = "dashed") +
#  labs(title = "Complementarity scores TAP-seq primer combinations",
#       color = "Complementarity") +
#  theme_bw()

# No primer pairs with complementarity scores above 47 were found, so they should be ok to use in
# multiplex PCRs.

# Export primers

# To finish the workflow, we can export the designed primers to data.frames, which can be written to
# .csv files for easy storage of primer sets.

# create data.frames for outer and inner primers
outer_primers_df <- primerDataFrame(best_outer_primers)
inner_primers_df <- primerDataFrame(best_inner_primers)

write.csv(outer_primers_df,'ipsc_best_outer_primers.csv')
write.csv(inner_primers_df,'ipsc_best_inner_primers.csv')

# the resulting data.frames contain all relevant primer data
head(outer_primers_df)

# Furthermore we can create BED tracks with the genomic coordinates of the primer binding sites for
# viewing in a genome browser.

# create BED tracks for outer and inner primers with custom colors
outer_primers_track <- createPrimerTrack(best_outer_primers, color = "steelblue3")
inner_primers_track <- createPrimerTrack(best_inner_primers, color = "goldenrod1")

# the output data.frames contain lines in BED format for every primer
head(outer_primers_track)
head(inner_primers_track)

# export tracks to .bed files ("" writes to the standard output, replace by a file name)
exportPrimerTrack(outer_primers_track, con = "./ipsc_outer_primers.bed")
exportPrimerTrack(inner_primers_track, con = "./ipsc_inner_primers.bed")

# Alternatively, primer tracks can be easily created for outer and inner primers and written to one
# .bed file, so that both primer sets can be viewed in the same track, but with different colors.
exportPrimerTrack(createPrimerTrack(best_outer_primers, color = "steelblue3"),
                  createPrimerTrack(best_inner_primers, color = "goldenrod1"),
                  con = "./ipsc_primers.bed")

# Session information

#All of the output in this vignette was produced under the following conditions:
sessionInfo()


```

